cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp

```diff
@@ -0,0 +1,389 @@
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int BAD_ARGUMENTS;
+}
+}
+
+using namespace DB;
+
+namespace local_engine
+{
+void SparkExchangeSink::consume(Chunk chunk)
+{
+    Stopwatch wall_time;
+    if (chunk.getNumRows() == 0)
+        return;
+    split_result.total_blocks += 1;
+    split_result.total_rows += chunk.getNumRows();
+    auto aggregate_info = chunk.getChunkInfos().get<AggregatedChunkInfo>();
+    auto input = inputs.front().getHeader().cloneWithColumns(chunk.detachColumns());
+    Stopwatch split_time_watch;
+    if (!sort_writer)
+        input = convertAggregateStateInBlock(input);
+    split_result.total_split_time += split_time_watch.elapsedNanoseconds();
+
+    Stopwatch compute_pid_time_watch;
+    PartitionInfo partition_info = partitioner->build(input);
+    split_result.total_compute_pid_time += compute_pid_time_watch.elapsedNanoseconds();
+
+    Block out_block;
+    for (size_t col_i = 0; col_i < output_header.columns(); ++col_i)
+    {
+        out_block.insert(input.getByPosition(output_columns_indicies[col_i]));
+    }
+    if (aggregate_info)
+    {
+        out_block.info.is_overflows = aggregate_info->is_overflows;
+        out_block.info.bucket_num = aggregate_info->bucket_num;
+    }
+    partition_writer->write(partition_info, out_block);
+    split_result.wall_time += wall_time.elapsedNanoseconds();
+}
+
+void SparkExchangeSink::onFinish()
+{
+    Stopwatch wall_time;
+    if (!dynamic_cast<LocalPartitionWriter*>(partition_writer.get()))
+    {
+        partition_writer->evictPartitions();
+    }
+    split_result.wall_time += wall_time.elapsedNanoseconds();
+}
+
+void SparkExchangeSink::initOutputHeader(const Block & block)
+{
+    if (!output_header)
+    {
+        if (output_columns_indicies.empty())
+        {
+            output_header = block.cloneEmpty();
+            for (size_t i = 0; i < block.columns(); ++i)
+                output_columns_indicies.push_back(i);
+        }
+        else
+        {
+            ColumnsWithTypeAndName cols;
+            for (const auto & index : output_columns_indicies)
+                cols.push_back(block.getByPosition(index));
+
+            output_header = Block(std::move(cols));
+        }
+    }
+}
+
+SparkExchangeManager::SparkExchangeManager(const Block& header, const String & short_name, const SplitOptions & options_, jobject rss_pusher): input_header(materializeBlock(header)), options(options_)
+{
+    if (rss_pusher)
+    {
+        GET_JNIENV(env)
+        jclass celeborn_partition_pusher_class =
+            CreateGlobalClassReference(env, "Lorg/apache/spark/shuffle/CelebornPartitionPusher;");
+        jmethodID celeborn_push_partition_data_method =
+            GetMethodID(env, celeborn_partition_pusher_class, "pushPartitionData", "(I[BI)I");
+        CLEAN_JNIENV
+        celeborn_client = std::make_unique<CelebornClient>(rss_pusher, celeborn_push_partition_data_method);
+        use_rss = true;
```

Comment:
could remove `use_rss`, just use `celeborn_client != nullptr`

Reply:
celeborn cannot bring performance improvement in parallel writing. In the case of rss, celeborn client will be moved to the Sink object, so an additional variable is needed